// posthog-js/lib/src/extensions/product-tours/components/ProductTourTooltip.js
"use strict";
var __read = (this && this.__read) || function (o, n) {
    var m = typeof Symbol === "function" && o[Symbol.iterator];
    if (!m) return o;
    var i = m.call(o), r, ar = [], e;
    try {
        while ((n === void 0 || n-- > 0) && !(r = i.next()).done) ar.push(r.value);
    }
    catch (error) { e = { error: error }; }
    finally {
        try {
            if (r && !r.done && (m = i["return"])) m.call(i);
        }
        finally { if (e) throw e.error; }
    }
    return ar;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ProductTourTooltip = ProductTourTooltip;
var jsx_runtime_1 = require("preact/jsx-runtime");
var hooks_1 = require("preact/hooks");
var product_tours_utils_1 = require("../product-tours-utils");
var utils_1 = require("../../../utils");
var globals_1 = require("../../../utils/globals");
var icons_1 = require("../../surveys/icons");
var window = globals_1.window;
function getOppositePosition(position) {
    var opposites = {
        top: 'bottom',
        bottom: 'top',
        left: 'right',
        right: 'left',
    };
    return opposites[position];
}
function scrollToElement(element, resolve) {
    var initialRect = element.getBoundingClientRect();
    var viewportHeight = window.innerHeight;
    var viewportWidth = window.innerWidth;
    var safeMarginY = viewportHeight / 6;
    var safeMarginX = viewportWidth / 6;
    var isInSafeZone = initialRect.top >= safeMarginY &&
        initialRect.bottom <= viewportHeight - safeMarginY &&
        initialRect.left >= safeMarginX &&
        initialRect.right <= viewportWidth - safeMarginX;
    if (isInSafeZone) {
        resolve();
        return;
    }
    element.scrollIntoView({ behavior: 'smooth', block: 'center' });
    var lastTop = initialRect.top;
    var stableCount = 0;
    var resolved = false;
    var checkStability = function () {
        if (resolved)
            return;
        var currentRect = element.getBoundingClientRect();
        if (Math.abs(currentRect.top - lastTop) < 1) {
            stableCount++;
            if (stableCount >= 3) {
                resolved = true;
                resolve();
                return;
            }
        }
        else {
            stableCount = 0;
        }
        lastTop = currentRect.top;
        setTimeout(checkStability, 50);
    };
    setTimeout(checkStability, 30);
    setTimeout(function () {
        if (!resolved) {
            resolved = true;
            resolve();
        }
    }, 500);
}
function ProductTourTooltip(_a) {
    var tour = _a.tour, step = _a.step, stepIndex = _a.stepIndex, totalSteps = _a.totalSteps, targetElement = _a.targetElement, onNext = _a.onNext, onPrevious = _a.onPrevious, onDismiss = _a.onDismiss;
    var appearance = (0, product_tours_utils_1.mergeAppearance)(tour.appearance);
    var _b = __read((0, hooks_1.useState)('initializing'), 2), transitionState = _b[0], setTransitionState = _b[1];
    var _c = __read((0, hooks_1.useState)(null), 2), position = _c[0], setPosition = _c[1];
    var _d = __read((0, hooks_1.useState)(null), 2), spotlightStyle = _d[0], setSpotlightStyle = _d[1];
    var _e = __read((0, hooks_1.useState)(step), 2), displayedStep = _e[0], setDisplayedStep = _e[1];
    var _f = __read((0, hooks_1.useState)(stepIndex), 2), displayedStepIndex = _f[0], setDisplayedStepIndex = _f[1];
    var previousStepRef = (0, hooks_1.useRef)(stepIndex);
    var isTransitioningRef = (0, hooks_1.useRef)(false);
    var isFirstRender = (0, hooks_1.useRef)(true);
    var isModalStep = !targetElement;
    var updatePosition = (0, hooks_1.useCallback)(function () {
        if (isModalStep) {
            return;
        }
        var rect = targetElement.getBoundingClientRect();
        setPosition((0, product_tours_utils_1.calculateTooltipPosition)(rect));
        setSpotlightStyle((0, product_tours_utils_1.getSpotlightStyle)(rect));
    }, [targetElement, isModalStep]);
    (0, hooks_1.useEffect)(function () {
        var isStepChange = previousStepRef.current !== stepIndex;
        var currentStepIndex = stepIndex;
        if (isFirstRender.current) {
            isFirstRender.current = false;
            previousStepRef.current = stepIndex;
            isTransitioningRef.current = true;
            if (isModalStep) {
                // Modal steps are just centered on screen - no positioning needed
                setTransitionState('visible');
                isTransitioningRef.current = false;
                return;
            }
            scrollToElement(targetElement, function () {
                if (previousStepRef.current !== currentStepIndex) {
                    return;
                }
                var rect = targetElement.getBoundingClientRect();
                setPosition((0, product_tours_utils_1.calculateTooltipPosition)(rect));
                setSpotlightStyle((0, product_tours_utils_1.getSpotlightStyle)(rect));
                setTransitionState('visible');
                isTransitioningRef.current = false;
            });
            return;
        }
        if (isStepChange) {
            previousStepRef.current = stepIndex;
            isTransitioningRef.current = true;
            setTransitionState('exiting');
            setTimeout(function () {
                if (previousStepRef.current !== currentStepIndex) {
                    return;
                }
                setDisplayedStep(step);
                setDisplayedStepIndex(stepIndex);
                setTransitionState('entering');
                if (isModalStep) {
                    // Modal steps don't need scrolling or position calculation
                    setTimeout(function () {
                        if (previousStepRef.current !== currentStepIndex) {
                            return;
                        }
                        setTransitionState('visible');
                        isTransitioningRef.current = false;
                    }, 50);
                    return;
                }
                scrollToElement(targetElement, function () {
                    if (previousStepRef.current !== currentStepIndex) {
                        return;
                    }
                    updatePosition();
                    setTimeout(function () {
                        if (previousStepRef.current !== currentStepIndex) {
                            return;
                        }
                        setTransitionState('visible');
                        isTransitioningRef.current = false;
                    }, 50);
                });
            }, 150);
        }
    }, [targetElement, stepIndex, step, updatePosition, isModalStep]);
    (0, hooks_1.useEffect)(function () {
        if (transitionState !== 'visible' || isModalStep) {
            return;
        }
        var handleUpdate = function () {
            if (!isTransitioningRef.current) {
                updatePosition();
            }
        };
        (0, utils_1.addEventListener)(window, 'scroll', handleUpdate, { capture: true });
        (0, utils_1.addEventListener)(window, 'resize', handleUpdate);
        return function () {
            window === null || window === void 0 ? void 0 : window.removeEventListener('scroll', handleUpdate, true);
            window === null || window === void 0 ? void 0 : window.removeEventListener('resize', handleUpdate);
        };
    }, [updatePosition, transitionState, isModalStep]);
    (0, hooks_1.useEffect)(function () {
        var handleKeyDown = function (e) {
            if (e.key === 'Escape') {
                onDismiss('escape_key');
            }
        };
        (0, utils_1.addEventListener)(window, 'keydown', handleKeyDown);
        return function () {
            window === null || window === void 0 ? void 0 : window.removeEventListener('keydown', handleKeyDown);
        };
    }, [onDismiss]);
    var handleOverlayClick = function (e) {
        e.stopPropagation();
        onDismiss('user_clicked_outside');
    };
    var handleTooltipClick = function (e) {
        e.stopPropagation();
    };
    var isLastStep = displayedStepIndex >= totalSteps - 1;
    var isFirstStep = displayedStepIndex === 0;
    var containerStyle = {
        '--ph-tour-background-color': appearance.backgroundColor,
        '--ph-tour-text-color': appearance.textColor,
        '--ph-tour-button-color': appearance.buttonColor,
        '--ph-tour-button-text-color': appearance.buttonTextColor,
        '--ph-tour-border-radius': "".concat(appearance.borderRadius, "px"),
        '--ph-tour-border-color': appearance.borderColor,
    };
    var isReady = isModalStep || (transitionState !== 'initializing' && position && spotlightStyle);
    var isVisible = transitionState === 'visible';
    if (!isReady) {
        return ((0, jsx_runtime_1.jsxs)("div", { class: "ph-tour-container", style: containerStyle, children: [(0, jsx_runtime_1.jsx)("div", { class: "ph-tour-click-overlay", onClick: handleOverlayClick }), (0, jsx_runtime_1.jsx)("div", { class: "ph-tour-spotlight", style: { top: '50%', left: '50%', width: '0px', height: '0px' } })] }));
    }
    // Modal step: centered on screen with overlay dimming, no spotlight/arrow
    if (isModalStep) {
        return ((0, jsx_runtime_1.jsxs)("div", { class: "ph-tour-container", style: containerStyle, children: [(0, jsx_runtime_1.jsx)("div", { class: "ph-tour-click-overlay", onClick: handleOverlayClick }), (0, jsx_runtime_1.jsx)("div", { class: "ph-tour-modal-overlay" }), (0, jsx_runtime_1.jsxs)("div", { class: "ph-tour-tooltip ph-tour-tooltip--modal", onClick: handleTooltipClick, children: [(0, jsx_runtime_1.jsx)("button", { class: "ph-tour-dismiss", onClick: function () { return onDismiss('user_clicked_skip'); }, "aria-label": "Close tour", children: icons_1.cancelSVG }), (0, jsx_runtime_1.jsx)("div", { class: "ph-tour-content", dangerouslySetInnerHTML: { __html: (0, product_tours_utils_1.renderTipTapContent)(displayedStep.content) } }), (0, jsx_runtime_1.jsxs)("div", { class: "ph-tour-footer", children: [(0, jsx_runtime_1.jsxs)("span", { class: "ph-tour-progress", children: [displayedStepIndex + 1, " of ", totalSteps] }), (0, jsx_runtime_1.jsxs)("div", { class: "ph-tour-buttons", children: [!isFirstStep && ((0, jsx_runtime_1.jsx)("button", { class: "ph-tour-button ph-tour-button--secondary", onClick: onPrevious, children: "Back" })), (0, jsx_runtime_1.jsx)("button", { class: "ph-tour-button ph-tour-button--primary", onClick: onNext, children: isLastStep ? 'Done' : 'Next' })] })] }), !appearance.whiteLabel && ((0, jsx_runtime_1.jsxs)("a", { href: "https://posthog.com/product-tours", target: "_blank", rel: "noopener noreferrer", class: "ph-tour-branding", children: ["Tour by ", icons_1.IconPosthogLogo] }))] })] }));
    }
    return ((0, jsx_runtime_1.jsxs)("div", { class: "ph-tour-container", style: containerStyle, children: [(0, jsx_runtime_1.jsx)("div", { class: "ph-tour-click-overlay", onClick: handleOverlayClick }), (0, jsx_runtime_1.jsx)("div", { class: "ph-tour-spotlight", style: isVisible && spotlightStyle
                    ? spotlightStyle
                    : {
                        top: '50%',
                        left: '50%',
                        width: '0px',
                        height: '0px',
                    } }), (0, jsx_runtime_1.jsxs)("div", { class: "ph-tour-tooltip ".concat(isVisible ? 'ph-tour-tooltip--visible' : 'ph-tour-tooltip--hidden'), style: {
                    top: "".concat(position.top, "px"),
                    left: "".concat(position.left, "px"),
                    opacity: isVisible ? 1 : 0,
                    transform: isVisible ? 'translateY(0)' : 'translateY(10px)',
                    transition: 'opacity 0.15s ease-out, transform 0.15s ease-out',
                }, onClick: handleTooltipClick, children: [(0, jsx_runtime_1.jsx)("div", { class: "ph-tour-arrow ph-tour-arrow--".concat(getOppositePosition(position.position)) }), (0, jsx_runtime_1.jsx)("button", { class: "ph-tour-dismiss", onClick: function () { return onDismiss('user_clicked_skip'); }, "aria-label": "Close tour", children: icons_1.cancelSVG }), (0, jsx_runtime_1.jsx)("div", { class: "ph-tour-content", dangerouslySetInnerHTML: { __html: (0, product_tours_utils_1.renderTipTapContent)(displayedStep.content) } }), (0, jsx_runtime_1.jsxs)("div", { class: "ph-tour-footer", children: [(0, jsx_runtime_1.jsxs)("span", { class: "ph-tour-progress", children: [displayedStepIndex + 1, " of ", totalSteps] }), (0, jsx_runtime_1.jsxs)("div", { class: "ph-tour-buttons", children: [!isFirstStep && ((0, jsx_runtime_1.jsx)("button", { class: "ph-tour-button ph-tour-button--secondary", onClick: onPrevious, children: "Back" })), (0, jsx_runtime_1.jsx)("button", { class: "ph-tour-button ph-tour-button--primary", onClick: onNext, children: isLastStep ? 'Done' : 'Next' })] })] }), !appearance.whiteLabel && ((0, jsx_runtime_1.jsxs)("a", { href: "https://posthog.com/product-tours", target: "_blank", rel: "noopener noreferrer", class: "ph-tour-branding", children: ["Tour by ", icons_1.IconPosthogLogo] }))] })] }));
}
